### PR TITLE
Remove template initialization from indexer-connector

### PR DESCRIPTION
**Pull request by MiguelazoDS**
|Related issue|
|---|
|#24351|

## Description (WIP)

This PR removes the template initialization from the indexer-connector, delegating the task to another module as described here https://github.com/wazuh/wazuh-indexer/issues/256

### Logs/Alerts example

<details><summary> Outdated </summary> 

Indexer deployed and initialized with template

```console
wazuh                                       [wazuh-alerts-4.x-*, wazuh-archives-4.x-*] 0 1
wazuh-states-vulnerabilities [wazuh-states-vulnerabilities]             1   []
```

Index created during index-connector initialization from template

```console
green open wazuh-states-vulnerabilities tRiOuybbQHqwjo-bRotNRw 1 0    0 0   208b   208b
```

If the template is not found the following message is displayed

```console
2024/07/17 19:13:34 indexer-connector[169554] indexerConnector.cpp:466 at operator()(): DEBUG: Unable to initialize IndexerConnector for index 'wazuh-states-vulnerabilities': HTTP response code said error (Status code: 404). Retrying in 2 seconds.
2024/07/17 19:13:34 indexer-connector[169554] indexerConnector.cpp:474 at operator()(): WARNING: IndexerConnector initialization failed for index 'wazuh-states-vulnerabilities', retrying until the connection is successful.
```

### Database generation (Github Action) 

Since the vulnerability scanner no longer fails due to a missing template parameter, the creation of the database works as expected.

```console
wazuh-modulesd:vulnerability-scanner:databaseFeedManager.hpp:337 operator() : Feed update process completed
wazuh-modulesd:content-updater:onDemandManager.cpp:169 stopServer : Server stopped
wazuh-modulesd:content-updater:action.hpp:138 stopActionScheduler : Scheduler stopped for 'vulnerability_feed_manager'
Error removing FD from interface.
+ rm -rf queue/indexer
+ rm -rf queue/sockets
+ rm -rf queue/router
+ rm -rf queue/vd_updater/tmp
+ rm -rf queue/vd/reports
+ rm -rf queue/vd/sync
+ rm -rf queue/vd/deltas
+ rm -rf queue/vd/state_track
+ rm -rf queue/keystore
+ VD_FILENAME=vd_1.0.0_vd_4.8.0.tar.xz
+ echo 'Compressing into '\''vd_1.0.0_vd_4.8.0.tar.xz'\'' ...'
Compressing into 'vd_1.0.0_vd_4.8.0.tar.xz' ...
+ tar -cJf vd_1.0.0_vd_4.8.0.tar.xz --owner=0 --group=0 --no-same-owner --no-same-permissions queue
+ '[' '!' -f vd_1.0.0_vd_4.8.0.tar.xz ']'
+ echo 'File '\''vd_1.0.0_vd_4.8.0.tar.xz'\'' generated successfully.'
File 'vd_1.0.0_vd_4.8.0.tar.xz' generated successfully.
++ du -h vd_1.0.0_vd_4.8.0.tar.xz
++ cut -f1
+ echo 'Size of '\''vd_1.0.0_vd_4.8.0.tar.xz'\'': 180M'
Size of 'vd_1.0.0_vd_4.8.0.tar.xz': 180M
root@jammy:/home/vagrant/wazuh# ls -l vd_1.0.0_vd_4.8.0.tar.xz
-rw-r--r-- 1 root root 188231144 Jul 17 18:58 vd_1.0.0_vd_4.8.0.tar.xz
```

### Update 

Final test 

Initial state indexer 
```console
root@jammy:/home/vagrant/wazuh# curl -k -u admin:admin https://127.0.0.1:9200/_cat/templates
wazuh-states-vulnerabilities [wazuh-states-vulnerabilities]             1   []
ss4o_metrics_template [ss4o_metrics-*-*]                         1 1 []
ss4o_traces_template  [ss4o_traces-*-*]                          1 1 []
root@jammy:/home/vagrant/wazuh# curl -k -u admin:admin https://127.0.0.1:9200/_cat/indices
green open .opensearch-observability   S8mhJ89nTWK4ZtIijv2xIA 1 0   0 0    208b    208b
green open .plugins-ml-config          3PSaegGxQ46UjQEgn1PSgA 1 0   1 0   3.9kb   3.9kb
green open .opendistro_security        Azbmbu8JReqiqXbBEnAVnw 1 0  10 0  65.1kb  65.1kb
```

Not undesired error or warning logs 

```console
root@jammy:/home/vagrant/wazuh# grep "ERROR" /var/ossec/logs/ossec.log
root@jammy:/home/vagrant/wazuh# grep "WARNING" /var/ossec/logs/ossec.log
2024/07/19 16:07:44 indexer-connector[222255] indexerConnector.cpp:82 at initConfiguration(): WARNING: No username and password found in the keystore, using default values.
```

Indexer connection successful.
```console
2024/07/19 16:07:45 indexer-connector[222255] indexerConnector.cpp:319 at initialize(): INFO: IndexerConnector initialized successfully for index: wazuh-states-vulnerabilities.
```

Index created 
```console
root@jammy:/home/vagrant/wazuh# curl -k -u admin:admin https://127.0.0.1:9200/_cat/indices
green open .opensearch-observability    S8mhJ89nTWK4ZtIijv2xIA 1 0   0 0    208b    208b
green open .plugins-ml-config           3PSaegGxQ46UjQEgn1PSgA 1 0   1 0   3.9kb   3.9kb
green open wazuh-states-vulnerabilities 90NaP-9EQxyG_kv2PXZ6SA 1 0  67 0 130.6kb 130.6kb
green open .opendistro_security         Azbmbu8JReqiqXbBEnAVnw 1 0  10 0  65.1kb  65.1kb
```

Vulnerabilities indexed and ossec.log file
[files.tar.gz](https://github.com/user-attachments/files/16316227/files.tar.gz)

</details> 

## Update 24/07

Indexer is expected to have template and index already initialized. 
```console
+ curl -k -u admin:admin -XGET https://127.0.0.1:9200/_cat/templates
wazuh-states-vulnerabilities [wazuh-states-vulnerabilities] 1   []
ss4o_metrics_template        [ss4o_metrics-*-*]             1 1 []
ss4o_traces_template         [ss4o_traces-*-*]              1 1 []
+ curl -k -u admin:admin -XGET https://127.0.0.1:9200/_cat/indices
green open .opensearch-observability    Y9k5YfxNQR6KG26SqrBoBQ 1 0  0 0   208b   208b
green open .plugins-ml-config           UT-uKaJVRPuCv77F5MXOTA 1 0  1 0  3.9kb  3.9kb
green open wazuh-states-vulnerabilities Qnh4b4x_QySI1fMJTZXIGw 1 0  0 0   208b   208b
green open .opendistro_security         4jzG9_kWQBacTUZZPwa2BA 1 0 10 0 65.1kb 65.1kb
```

If we remove the index we can see some warning messages 

```console
grep -E "ERR.*|WARN.*|CRIT.*" ossec.log
2024/07/24 15:48:52 indexer-connector: WARNING: No username and password found in the keystore, using default values.
2024/07/24 15:48:52 wazuh-modulesd:vulnerability-scanner: ERROR: Error opening the database: Couldn't find column family: 'vendor_map', trying to re-download the feed.
2024/07/24 16:16:06 indexer-connector[26512] indexerConnector.cpp:82 at initConfiguration(): WARNING: No username and password found in the keystore, using default values.
2024/07/24 16:21:48 indexer-connector[29690] indexerConnector.cpp:82 at initConfiguration(): WARNING: No username and password found in the keystore, using default values.
2024/07/24 16:21:48 indexer-connector[29690] indexerConnector.cpp:390 at operator()(): WARNING: Failed to sync agent '000' with the indexer.
2024/07/24 16:22:46 indexer-connector[31860] indexerConnector.cpp:82 at initConfiguration(): WARNING: No username and password found in the keystore, using default values.
2024/07/24 16:22:47 indexer-connector[31860] indexerConnector.cpp:390 at operator()(): WARNING: Failed to sync agent '000' with the indexer.
```

Files
[files.tar.gz](https://github.com/user-attachments/files/16366614/files.tar.gz)

### CODE REVIEW

**Comment by GabrielEValenzuela** [comment]
# Review I

## Code Review 🧑‍💻

### Comments and Suggestions
- **General Feedback:**  
  The code quality is clear and well-formed.

- **Change requests:**  
  - Nonde

### Conclusion
- **Approval Status:**  
  - [X] Approved  
  - [ ] Needs Changes  
  - [ ] Rejected

---

## Test Review 🧪


### Test Plan
- **Test Cases:**  
- Installing a fresh install of this branch and connecting to the Wazuh ecosystem.

- **Test Environment:**  
OS Ubuntu 24
- **Test Steps:**  

1. Install Wazuh-indexer version 4.8.1
2. Install Wazuh-dashboard version 4.8.1
3. Install Filebeat and configurable [following](https://documentation.wazuh.com/current/installation-guide/wazuh-server/step-by-step.html)
4. Compile the branch and start the manager once installed.

### Test Results

The indexer connection wasn't successful due to the missing index inside of the indexer

<details><summary>Screenshots</summary>

image


image


image

image


</details>


### Conclusion
- **Overall Test Result:**  
  - [ ] Passed  
  - [ ] Passed with Issues  
  - [X] Failed

> [!NOTE]
> The missing index doesn't mean that the current logic is bad, it could related to the indexer configuration. The creation of a new index requires a mapping and that information is not available, if the mapping exists the test can run again and check the result. The upgrade from `4.8.x` to `5.0.0` has a lot of incompatibilities, for example, the keystoring and the test couldn't be done.
---

**Review by GabrielEValenzuela — COMMENTED** [review]
[Report](https://github.com/wazuh/wazuh/pull/24654#issuecomment-2243497906)

**Review by GabrielEValenzuela — APPROVED** [review]

File: src/shared_modules/indexer_connector/src/indexerConnector.cpp [hunk 1]
```diff
@@ -292,36 +291,36 @@ void IndexerConnector::initialize(const nlohmann::json& templateData,
     };
 
     // Define the success callback
-    auto onSuccess = [](const std::string&)
+    nlohmann::json templateData {};
+    auto onSuccess = [&templateData](const std::string& response)
     {
-        // Not used
+        nlohmann::json jsonResponse = nlohmann::json::parse(response);
+        templateData = jsonResponse.at("index_templates").front().at("index_template").at("template");
```
Comment (sebasfalcone) [inline src/shared_modules/indexer_connector/src/indexerConnector.cpp:298]: We can avoid the use of the extra variable
```suggestion
        templateData = nlohmann::json::parse(response).at("index_templates").front().at("index_template").at("template");
```

Also, we should catch the possible exceptions here to improve the error messages down the line

```suggestion
        try
        {
            templateData = nlohmann::json::parse(response).at("index_templates").front().at("index_template").at("template");
        }
        catch (const nlohmann::json::exception& e)
        {
            logDebug1(IC_NAME, "Failed to parse index template: %s", e.what());
            throw std::runtime_error("Failed to parse index template.");
        }
```

File: src/shared_modules/indexer_connector/tests/component/indexerConnector_test.cpp
Comment (sebasfalcone) [inline src/shared_modules/indexer_connector/tests/component/indexerConnector_test.cpp:1]: Lets add a test when an invalid response is obtained (missing template or something like that)

**Review by sebasfalcone — COMMENTED** [review]
Just two minor details and we are ready to go

File: src/shared_modules/indexer_connector/tests/component/indexerConnector_test.cpp
Comment (MiguelazoDS) [reply src/shared_modules/indexer_connector/tests/component/indexerConnector_test.cpp:1]: Added

**Review by sebasfalcone — APPROVED** [review]
LGTM!

**Review by GabrielEValenzuela — APPROVED** [review]
Great job!

File: src/shared_modules/indexer_connector/include/indexerConnector.hpp [hunk 2]
```diff
@@ -54,16 +54,12 @@ class EXPORTED IndexerConnector final
...
      *
-     * @param templateData Template data.
-     * @param indexName Index name.
      * @param selector Server selector.
      * @param secureCommunication Secure communication.
      */
-    void initialize(const nlohmann::json& templateData,
-                    const std::shared_ptr<ServerSelector>& selector,
-                    const SecureCommunication& secureCommunication);
+    void initialize(const std::shared_ptr<ServerSelector>& selector, const SecureCommunication& secureCommunication);
```
Comment (Dwordcito) [inline src/shared_modules/indexer_connector/include/indexerConnector.hpp:62]: having this in the interface doesn't make sense.

File: src/shared_modules/indexer_connector/src/indexerConnector.cpp [hunk 3]
```diff
@@ -292,36 +291,44 @@ void IndexerConnector::initialize(const nlohmann::json& templateData,
...
-                                onSuccess,
-                                onError,
-                                "",
-                                DEFAULT_HEADERS,
-                                secureCommunication);
-
     // Initialize Index.
-    HTTPRequest::instance().put(HttpURL(selector->getNext() + "/" + m_indexName),
-                                templateData.at("template"),
+    HTTPRequest::instance().get(HttpURL(selector->getNext() + "/_index_template/" + m_indexName),
```
Comment (Dwordcito) [inline src/shared_modules/indexer_connector/src/indexerConnector.cpp:310]: This must also be removed. We do not initialize anything else here. Remove this method.

File: src/shared_modules/indexer_connector/src/indexerConnector.cpp [hunk 4]
```diff
@@ -452,7 +451,7 @@ IndexerConnector::IndexerConnector(
 
     m_initializeThread = std::thread(
```
Comment (Dwordcito) [inline src/shared_modules/indexer_connector/src/indexerConnector.cpp:452]: remove this thread.

**Review by Dwordcito — CHANGES_REQUESTED** [review]

File: src/shared_modules/indexer_connector/include/indexerConnector.hpp (hunk 2 above)
Comment (MiguelazoDS) [reply src/shared_modules/indexer_connector/include/indexerConnector.hpp]: Yeap, this was called by the removed thread, removed.

File: src/shared_modules/indexer_connector/src/indexerConnector.cpp (hunk 3 above)
Comment (MiguelazoDS) [reply src/shared_modules/indexer_connector/src/indexerConnector.cpp]: Wrong approach, removed. Template and index are supposed to be initialized.

File: src/shared_modules/indexer_connector/src/indexerConnector.cpp (hunk 4 above)
Comment (MiguelazoDS) [reply src/shared_modules/indexer_connector/src/indexerConnector.cpp]: Removed

**Review by Dwordcito — APPROVED** [review]

**Pull request merged**